halo2_proofs/src/plonk/permutation/keygen.rs

```diff
@@ -240,13 +248,311 @@ impl Assembly {
+
+        // If left and right are in the same cycle, do nothing.
+        if (left_cycle == right_cycle) && (left_cycle.is_some()) && (right_cycle.is_some()) {
+            return Ok(());
+        }
+
+        let left_cycle_size = match left_cycle {
+            // this should unwrap safely
+            Some(i) => self.cycles[*i].len(),
+            None => 1,
+        };
+
+        let right_cycle_size = match right_cycle {
+            // this should unwrap safely
+            Some(i) => self.cycles[*i].len(),
+            None => 1,
+        };
+
+        if left_cycle_size < right_cycle_size {
+            std::mem::swap(&mut left_cycle, &mut right_cycle);
+            std::mem::swap(&mut left_column, &mut right_column);
+            std::mem::swap(&mut left_row, &mut right_row);
+        }
+
+        // extract cycle elements
+        let right_cycle_elems = match right_cycle {
+            Some(i) => {
+                let entry = self.cycles[*i].clone();
+                self.cycles[*i] = BTreeSet::new();
+                entry
+            }
+            None => [(right_column, right_row)].into(),
+        };
+
+        assert!(right_cycle_elems.contains(&(right_column, right_row)));
+
+        // merge cycles
+        let cycle_idx = match left_cycle {
+            Some(i) => {
+                let entry = &mut self.cycles[*i];
+                entry.extend(right_cycle_elems.clone());
+                *i
+            }
+            // if they were singletons -- create a new cycle entry
+            None => {
+                let mut set: BTreeSet<(usize, usize)> = right_cycle_elems.clone();
+                set.insert((left_column, left_row));
+                self.cycles.push(set);
+                let cycle_idx = self.cycles.len() - 1;
+                self.aux.insert((left_column, left_row), cycle_idx);
+                cycle_idx
+            }
+        };
+
+        let index_updates = vec![cycle_idx; right_cycle_elems.len()].into_iter();
+        let updates = right_cycle_elems.into_iter().zip(index_updates);
+
+        self.aux.extend(updates);
+
+        Ok(())
+    }
+
+    fn mapping_at_idx(&self, col: usize, row: usize) -> (usize, usize) {
+        if let Some(cycle_idx) = self.aux.get(&(col, row)) {
+            let cycle = &self.cycles[*cycle_idx];
+            let mut cycle_iter = cycle.range((
+                std::ops::Bound::Excluded((col, row)),
+                std::ops::Bound::Unbounded,
+            ));
+            // point to the next node in the cycle
+            match cycle_iter.next() {
+                Some((i, j)) => (*i, *j),
+                // wrap back around to the first element which SHOULD exist
+                None => *(cycle.iter().next().unwrap()),
+            }
+        // is a singleton
+        } else {
+            (col, row)
+        }
+    }
+
+    pub(crate) fn build_vk<'params, C: CurveAffine, P: Params<'params, C>>(
```

Comment:
Perhaps we could extract the body of this function out for sharing, the difference of these 2 impl is only the mapping function, so we could do something like:

```rust
pub(crate) fn build_pk<'params, C: CurveAffine, P: Params<'params, C>>(
    params: &P,
    domain: &EvaluationDomain<C::Scalar>,
    p: &Argument,
    mapping: impl Fn(usize, usize) -> (usize, usize) + Sync,
) -> ProvingKey<C> {
    // ...
}
```

And for the first impl we do:

```rust
    pub(crate) fn build_pk<'params, C: CurveAffine, P: Params<'params, C>>(
        self,
        params: &P,
        domain: &EvaluationDomain<C::Scalar>,
        p: &Argument,
    ) -> ProvingKey<C> {
        build_pk(params, domain, p, |i, j| self.mapping[i][j])
    }
```

For the second impl we do:

```rust
    pub(crate) fn build_pk<'params, C: CurveAffine, P: Params<'params, C>>(
        self,
        params: &P,
        domain: &EvaluationDomain<C::Scalar>,
        p: &Argument,
    ) -> ProvingKey<C> {
        build_pk(params, domain, p, |i, j| self.mapping_at_idx(i, j))
    }
```